### ISSUE TITLE
[Feature request] Data/Control channel control

### ISSUE THREAD

**Issue by juztas** [xrootd/xrootd#1192]
It would be nice to have a way to control the data_interface and control_interface for the client streaming and third party copy (similarly like GridFTP does). https://gridcf.org/gct-docs/latest/gridftp/admin/index.html#_data_interface 

One of the features I miss moving away from GridFTP. Would it be possible to have such implementation and or how hard would be to have this?

**Comment by bbockelm** [comment]
Hi @juztas - this exists already, no?

That is, both Xrootd-TPC and HTTP-TPC use a separate connection(s) for data and control channel.

In case of HTTP-TPC, I know the dCache team has demonstrated pushing the control channel over HTTP and using GridFTP for the data channel(s).

**Comment by juztas** [comment]
Is this what admins can control? Do we have documentation for this? I was going back and forth and don't see any such ways. The idea is: Does not matter which protocol (ROOT, HTTP, DAV) and for xrootd server I want to configure:
the control interface is 192.168.0.1:2094 and the data interface is 10.0.0.1. Any client connecting to 192.168.0.1:2094 and want to transfer file locally or to another xrootd server would use 10.0.0.1 interface/network. So Client/Server Auth/mappinng happens via 192.168.0.1:2094, but data flows via 10.0.0.1.

**Comment by abh3** [comment]
There is no way of doing that. We never found a reason as we can flow the 
data throug the same IP address and avoid the standard overhead (not that 
it matters much if you're transfering a big file). Also, how would your 
proposal be applicable for anything but inside transfers (i.e intranet)?

**Comment by bbockelm** [comment]
@juztas - in the end, HTTP-TPC is "simply" calling `libcurl` to do the heavy lifting while Xrootd-TPC is invoking an arbitrary executable provided by the sysadmin.

In the case of HTTP-TPC, the best way to do this is through providing a unique source URL that forces things through the desired path.  For Xrootd-TPC -- well, the sky's the limit.

Looking at `libcurl`'s documentation, we could provide a bind interface:

https://curl.haxx.se/libcurl/c/CURLOPT_INTERFACE.html

Would that help here?

**Comment by juztas** [comment]
Everything is possible, no? Maybe my explanation was too minimalistic. Few more examples
1. We have direct links (a private network) to a few HPC/Other clusters locally.  Other clusters are connected to outside world (public network) at 10gbit or only 1gbit and we cant overload these links, while the private network link we have is min 40gbit or 100gbit between them. I could always use the private IP to access data on Tier2 from another cluster, but I have to enforce the use of `XrdSecGSISRVNAMES=.....` to bypass validation. It works, but what I want that any control communication continues to be via a public interface, while data is transferred via private.
2. Third-party copy - (Xrootd-TPC) I could achieve the same with XrdSecGSI... env hack - but it does not look something production-ready (it's a hack and I need to keep lists of DNs and what where to whitelist). So for 3rd party copy like GridFTP does would be awesome. @bbockelm for http-tpc - looks like that would work, leave communication via public, but data interface via private. Not sure how much changes it would require as two end host servers need to agree to send/receive data via private

**Comment by bbockelm** [comment]
For direct streaming / (1) - so what you are asking is that the client contacts the server on (public) interface A (and not private interface B!) but then download all the files over private interface B?  I suppose that a routing / encapsulation-based approach is not feasible?

For TPC / (2) - how does this work for GridFTP?  IIUC, this would cause GridFTP to always route data over the private interface (regardless of who the Caltech GridFTP server is talking with).  Is that acceptable?

**Comment by abh3** [comment]
So, please be patient with me as I try to fully understand the picture. I went through a related discussion with GSI some time ago. They had a similar issue and it took a while to understand the network topology before we could solve their data routing requirements without inflicting pain on people who didn't have their issue. The end result was an additional configuration option that described the topology so that xrootd could choose the appropriate interface. Most sites don't need to do that so it worked out fine; they just ignore the option.

So, in the xrootd case, data channels are established using the same IP address that was used for the control channel.  That isn't a requirement. The only requirement is that the data connection is made to the same host as the control connection because data follows control. So, is it the case that the initial login will always be to the host sending or receiving data? If so, the only thing you are requesting is that a different interface be used for the actual data. Is that correct?

Next, so assume that data and control will always be to the same host. Is it the case that the two interfaces are visible via DNS? That's not always the case, so that is why I ask. It's not a requirement but it makes life easier and a tad more secure if that is the case.

Once I get the above information, there are likely several ways this can be accomplished without disrupting the common case scenario. For instance, one possibility is for an xrootd server to advertise the preferred data path so that data channels can use it; if it's accessible from the client's perspective. We did that some time ago for different reasons but no one used the feature so we killed it. We can always bring it back. Another, perhaps simpler, solution is to redirect the data channel to the interface you would prefer to use for data.

**Comment by juztas** [comment]
Brian  - 1 - yes, routing is not an option for most cases. 2 - yes, GridFTP will always keep routing data via private, while auth and 3rd party client keeps getting transfer stat updates via public. (For me - yes, acceptable. Maybe there are more use cases?)

Andrew - If hosts behind NAT, not an issue, then yes - the host making a connection to the control channel will be the same asking for data channel.
For DNS - Usually no, there are no entries in DNS of our servers - So it will not resolve.

**Comment by abh3** [comment]
Hi Justas,

So, would this work for you? For a server you simply specify the IP 
address to use for the data channels. The server will simply redirect the 
client to use that IP address for each data channel it creates. Now, that 
does mean that the client needs to be able to connect using that IP 
address. At the moment, I don't see any way of verifying that would be 
true in the general case but I'll work on that part.

Andy

**Comment by juztas** [comment]
Hi, yes - that would work. Thanks!

**Comment by xrootd-dev** [comment]
Hi Justas,

You might just be able to test this right now if you don't mind the 
control channel going to the interface where data channels go. Look at the 
following directive:

https://xrootd.slac.stanford.edu/doc/dev50/xrd_config.htm#_Toc39615280

you can create a static redirect (relatively flexible). So, if you 
redirect the control channel to the interface where you want the data to 
go, the data channels will follow.

Andy

**Comment by juztas** [comment]
If I read it correctly - this is only for metadata operations, correct? So it would fail authentication. I tried a few options and it seems to redirect everything, even auth.  
```
credentials Secgsi: ErrParseBuffer: server certificate CN 'transfer-........' does not match the expected format(s): '[*/]172.31.14.11[/*]' (default); exceptions are controlled by the env XrdSecGSISRVNAMES: kXGS_cert
[2020-05-19 16:48:04.086265 -0700][Error  ][AsyncSock         ] [172.31.14.11:3609 #0.0] Socket error while handshaking: [FATAL] Auth failed
[2020-05-19 16:48:04.086272 -0700][Debug  ][AsyncSock         ] [172.31.14.11:3609 #0.0] Closing the socket
```

**Comment by abh3** [comment]
You are correct! I forgot about that not so little detail. OK, so now we 
know what needs to be done.

**Comment by abh3** [comment]
Just an update. We are working on a simple solution but it seems simple is not in the cards. Whatever the solution, it likely requires that you use an R5 client. It cuts in two ways, either we update the client to effect a split data path or we require the client to use TLS to effect the split data path (the latter only requires server-side changes). The data path doesn't need to use TLS, we just need TLS to securely do the split from the control channel. However, TLS is only available in R5. So, either way we are stuck to requiring R5 on both ends. What I mean by R5 is that the server and client be at release 5. We don't see any way we can make this backward compatible other than the standard fallback of using the slow connection for the data (which we would do, by default, for backward compatibility). Is that going to be a problem for you? Would you like at least the option if the client is not R5 to either use the slow path or just terminate the request?

**Comment by abh3** [comment]
BTW if you don't care about security and MIM attacks we could do this without TLS but I wouldn't recommend it.

**Comment by abh3** [comment]
I hope you are still interested in this as we spent some amount of time nailing down exactly how this could be done in a backwardly compatible way. Of course, other priorities interfered so it took some time. The solution is to relay bind preferences pre-login time. To that end we plan to implement the directive:

xrootd.bindpref <interface> [<interface> [...]]
where <interface> is a hostname:port, ipv4addr:port, or [ipv6addr]:port

The server will accept bind requests on all of the listed interfaces. The client will preferentially establish data channels on these interfaces. If more than one exists it will round-robin across all of them. 

I think that meets your requirements, yes?

**Comment by simonmichal** [comment]
This has been implemented in `5.4.0`.
